Archipelago.MultiClient.Net/Helpers/LocationCheckHelper.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text;
+using Archipelago.MultiClient.Net.Cache;
+using Archipelago.MultiClient.Net.Enums;
+using Archipelago.MultiClient.Net.Models;
+using Archipelago.MultiClient.Net.Packets;
+
+namespace Archipelago.MultiClient.Net.Helpers
+{
+    public class LocationCheckHelper
+    {
+        private readonly List<int> locationsChecked = new List<int>();
+        private readonly ArchipelagoSocketHelper socket;
+        private object locationsCheckedLockObject = new object();
+        
+        private bool awaitingLocationInfoPacket;
+        private Action<LocationInfoPacket> locationInfoPacketCallback;
+
+        public ReadOnlyCollection<int> AllLocationsChecked => new ReadOnlyCollection<int>(locationsChecked);
+
+        internal LocationCheckHelper(ArchipelagoSocketHelper socket)
+        {
+            this.socket = socket;
+
+            socket.PacketReceived += Socket_PacketReceived;
+        }
+
+        /// <summary>
+        ///     Submit the provided location ids as checked locations.
+        /// </summary>
+        /// <param name="ids">
+        ///     Location ids which have been checked.
+        /// </param>
+        public void CompleteLocationCheck(params int[] ids)
+        {
+            lock (locationsCheckedLockObject)
+            {
+                locationsChecked.AddRange(ids);
+                socket.SendPacket(new LocationChecksPacket()
+                {
+                    Locations = ids.ToList()
+                });
+            }
+        }
+
+        /// <summary>
+        ///     Ask the server for the items which are present in the provided location ids.
+        /// </summary>
+        /// <param name="callback">
+        ///     An action to run once the server responds to the scout packet.
+        ///     If the argument to the action is null then the server responded with an InvalidPacket response.
+        /// </param>
+        /// <param name="ids">
+        ///     The locations ids which are to be scouted.
+        /// </param>
+        public void ScoutLocations(Action<LocationInfoPacket> callback = null, params int[] ids)
```

Comment:
I think this is not like any of the async patterns we discussed before, The issue i see with this is that the callback is not tied to scope i can for example do this
ScoutLocations(p => ScoutUnderWaterChecks(p), 1337146, 1337145, 1337149);
ScoutLocations(p => ScoutLibraryChecks(p), 1337098, 1337099);

This will case ScoutLibraryChecks to be called with the results of all those 5 items and ScoutUnderWaterChecks never called. this implementation detail is not clear to the caller from the outside

Reply:
Let me test this out. I don't think you're right in the problem you're pointing out, but it does give me an indication of another issue.

Reply:
It can be an issue to call ScoutLocations() multiple times, the last call will overwrite the callback. Since the packets are responded to by the server in order, you could end up with odd behavior.

My suggestion to resolve this is: scout all locations at once. Or, use the following model of chaining calls:
```cs
void ScoutAllLocations(LocationCheckPacket p)
{
    // ...
    ScoutLocations(p => ScoutLibraryChecks(p), 1337098, 1337099);
}
ScoutLocations(p => ScoutAllLocations(p), 1337146, 1337145, 1337149);
```